src/track/track.cpp

```diff
@@ -1276,6 +1276,37 @@ void Track::importPendingCueInfosMarkDirtyAndUnlock(
     emit cuesUpdated();
 }
 
+QMap<int, MacroPointer> Track::getMacros() const {
+    const auto locked = lockMutex(&m_qMutex);
+    return m_macros;
+}
+
+void Track::setMacros(const QMap<int, MacroPointer>& macros) {
+    const auto locked = lockMutex(&m_qMutex);
+    m_macros = macros;
+}
+
+void Track::addMacro(int slot, const MacroPointer& pMacro) {
+    auto locked = lockMutex(&m_qMutex);
+    m_macros.insert(slot, pMacro);
+    if (pMacro->isDirty()) {
+        markDirtyAndUnlock(&locked);
+    }
+}
+
+bool Track::isDirty() {
+    const auto locked = lockMutex(&m_qMutex);
+    if (m_bDirty) {
+        return true;
+    }
+    for (const MacroPointer& pMacro : qAsConst(m_macros)) {
```

Comment:
So the dirty flag no longer indicates that something has changed.

Instead it has to be set in setMacros() and the original function should remain unchanged.

The dirty flag is a questionable optimization (accepting false positives), but we need to keep its original meaning to avoid regressions.

Reply:
Sure, this is not meant to stay. But can we address this in a follow-up PR? I am afraid of further conflicts if we push this off even longer.

Reply:
Noted on https://github.com/mixxxdj/mixxx/pull/2989
It can't be set in `setMacros` only, as they are mutable.

Reply:
Then I expect at least a FIXME comment that explains the intention and how it should be resolved.

Nevertheless, I suggest to avoid introducing ugly hacks in the first place. Who is supposed to fix this and when?

Reply:
I will fix this. This will need follow-up work anyways, as I will be using and polishing the features. But I doubt that I will get this done soon enough to warrant delaying this PR. But I can add a FIXME comment, yes.

Reply:
Please also restore the original ordering. This will help to resolve conflicts later, see #4534.